### DISCUSSION TITLE
Is it possible to save and restore the status of yazi?

### DISCUSSION THREAD

**Discussion by MasouShizuka** [sxyazi/yazi#965] [1. Q&A]
Is it possible to save the current status when exiting yazi, and read and restore it when starting?

For example:
There may be three tabs at that time, each tab is in a different path, and some tabs may have changed the sorting method, etc.
Save the current status to `~/.yazi_session` when exiting.

Afterwards, `~/.yazi_session` is read during startup, and the status of these three tabs can be restored later.

Currently, is there any way to achieve this feature?

**Comment by sxyazi** [comment]
DDS is designed to achieve this, see the docs at https://yazi-rs.github.io/docs/dds

You can find more examples about DDS on the v0.2.5 release page, https://github.com/sxyazi/yazi/releases/tag/v0.2.5

**Comment by MasouShizuka** [comment]
I also read the content of `DDS`, but I only know that it can synchronize some data across instances, and I still don’t understand how to save and restore the state.

Could you please give a simple example?

**Comment by MasouShizuka** [comment]
I read the content of `DDS` carefully and roughly realized that it is an event feature that provides registration events and actions when receiving events.

But it seems that `DDS` cannot implement the function of saving and reading the state of `yazi`.

Currently `DDS` can output specified events to `stdout`, which means that the events can be saved locally.
To achieve the function of restoring the state, each line of the file must be read and the event and content must be sent to `yazi`.
But this might not be a good idea because:
1. Need enough events
2. Restoring the state requires executing events for each line of the file, which is very time-consuming.
3. Saved File will grow endlessly

So, I thought if it is possible to provide a feature to initialize `yazi` through configuration.
To put it simply, can save each `tab` and the path of `tab`.
In this way, reading the file can restore the specified environment.

**Comment by sxyazi** [comment]
I think DDS currently supports the functionality you mentioned. 

You can bind a `plugin` command for creating tabs (`t` key), in which you can get the paths of all tabs through [`cx.tabs`](https://yazi-rs.github.io/docs/plugins/types#app-data) and store the tab status as a static message permanently through [`ps.pub_static`](https://yazi-rs.github.io/docs/plugins/utils/#ps.pub_static):

```
{ on = [ "t" ], run = [ "tab_create --current" , "plugin --sync your-tab-sync-plugin"] }
```

so that you will receive this event again the next time Yazi starts through [`ps.sub_remote()`](https://yazi-rs.github.io/docs/plugins/utils/#ps.sub_remote), and restores each tab by calling [`ya.manager_emit()`](https://yazi-rs.github.io/docs/plugins/utils/#ya.manager_emit).

This is an example of how the `bookmarks.yazi` plugin saves and restores bookmarks added by users through DDS, https://github.com/dedukun/bookmarks.yazi/blob/cc141d8ef81ab92385a576bab2cc7d6d5adeaeda/init.lua#L46

**Comment by MasouShizuka** [comment]
Oh, I totally didn't expect `pub_static` to be able to store data, even after exiting.

But after a rough search, I couldn't figure out where `pub_static` stores the data.
Could you please tell me how it works?

**Comment by MasouShizuka** [comment]
OK, I wrote this function as a plugin, here is its repository [projects.yazi](https://github.com/MasouShizuka/projects.yazi).
Welcome to use it and promote it:)

By the way, I want to do something before exiting `yazi`, but there seems to be no good way.

`DDS`'s `builtin kinds` does not seem to include `quit`, so `sub()` is not possible.

I also thought about implementing it through a key:
```lua
local quit_and_save_last = ya.sync(function(state)
    local projects = _get_projects()
    projects.last = _get_current_project()

    _save_state(projects)

    ya.manager_emit("quit", {})
end)
```
This can be done, but a window will pop up before quit: prompting that there is still 1 task that has not been completed.

So, is there any way to implement it?

**Reply by sxyazi** [reply]
> OK, I wrote this function as a plugin, here is its repository [projects.yazi](https://github.com/MasouShizuka/projects.yazi).

Awesome, great work!

> I also thought about implementing it through a key:

Yeah, this is the currently recommended approach. After handling the state saving in the plugin, send a `quit` to Yazi to exit. 

I think the popup is necessary because if the user is in the middle of a copy or cut operation, exiting abruptly could be dangerous and could lead to data loss if files are only partially copied. 

If what's blocking you is the preloading rather than copy or cut operations, then you can set [`suppress_preload`](https://yazi-rs.github.io/docs/configuration/yazi/#tasks.suppress_preload) to permanently disable this prompt. Perhaps we could also add a new flag to `quit` command to ignore preloading concerns.

**Reply by MasouShizuka** [reply]
No, even when `suppress_preload = true` is set in `yazi.toml`, pressing `q` to execute `quit_and_save_last()` still displays `1 task running, sure to quit? (y/N)`.
When pressing `esc` to cancel, the task list is empty and I can't see what the task is.

At that time, no other operations were performed, including copying, cutting, etc., so I thought this `task` was the process of saving the project.

So I tried to add `ya.sleep(0.5)` above `ya.manager_emit("quit", {})` to wait for the task to complete, but after adding it, the task could not be completed. I checked and found an error:
```
Micro plugin failed:
Failed to execute sync block-11 in `projects` plugin
stack traceback:
        [C]: in local 'poll'
        [string "?"]:4: in upvalue 'quit_and_save_last'
        [string "D:\Downloads\yazi\yazi-plugin\src\isolate\ent..."]:268: in function <[string "D:\Downloads\yazi\yazi-plugin\src\isolate\ent..."]:229>
```

**Reply by sxyazi** [reply]
I think this is because at the moment when the current plugin sends the quit request, the current plugin is still running, which leads to this confirmation prompt. Would you like to file an issue with that?

**Reply by MasouShizuka** [reply]
Ok, I'll open an issue for this.
